README.md

[![NPM version][npm-version-image]][npm-url]
[![NPM downloads][npm-downloads-image]][npm-url]
[![MIT License][license-image]][license-url]
[![Build Status: Linux][travis-image]][travis-url]
[![Build Status: Windows][appveyor-image]][appveyor-url]
[![Coverage Status][coveralls-image]][coveralls-url]


# Time checker

  v.0.1.2

  Checks current time in a specified period


## Install

`npm install time-checker`


## Usage Example

  Check if current time is my working days free time? Working time 10:00-18:00, Saturday and Sunday are a days off. Result depends on current time.

```javascript
var CheckTime = require('time-checker');
var checkTime = new CheckTime({ daysOff : [0, 6] });
checkTime.byConditions(['18:00-23:59', '00:00-10:00', 'workdays']);
```


## Option parameters
  - **daysOff** - list of days off. Array of numbers (`0` - Sun, `1` - Mon, ..., `6` - Sat)
  - **holidays** - list of holiday dates. Array of string (`'14.10'` means 14'th of October of each year)
  - **timeZone** - zone name used in value of TZ environment variable (ex. `America/Chicago`)


## Check time by conditions
  `checkTime.byConditions(<conditions>, [<current timestamp>])`

  - **conditions** - string or array of the strings with time conditions:
    - time interval (ex. `'9:00-21:00'`)
    - type of the day (`'workdays'` or `'restdays'`)
  - **current timestamp** - Unix Timestamp in milliseconds (ex. `1465200806000`), not required. If not set, then uses current timestamp.


## More Examples

  Set Saturday and Sunday as a days off. Holidays are first and seventh of January.

```javascript
var CheckTime = require('time-checker');
var checkTime = new CheckTime({ daysOff : [0, 6], holidays : ['1.01', '7.01'], timeZone: 'Etc/GMT' });

// checking 'Mon Jun 06 11:13:26 2016' on perioud '9:00-17:00'
checkTime.byConditions('9:00-17:00', 1465200806000); // true

// is Sunday is a rest day?
checkTime.byConditions('restdays', 1465200806000); // true

// is 1'st of January is a rest day?
checkTime.byConditions('restdays', 1483272000000); // true

// is 'Mon Jun 06 11:13:26 2016' is a rest day with perioud '9:00-17:00'?
checkTime.byConditions(['9:00-17:00', 'restdays'], 1465200806000); // false
```


## Change Log

  [all changes](CHANGELOG.md)


## Created by

  Dimitry Ivanov <2@ivanoff.org.ua> # curl -A cv ivanoff.org.ua

[license-image]: http://img.shields.io/badge/license-MIT-blue.svg?style=flat
[license-url]: LICENSE

[npm-url]: https://npmjs.org/package/time-checker
[npm-version-image]: http://img.shields.io/npm/v/time-checker.svg?style=flat
[npm-downloads-image]: http://img.shields.io/npm/dm/time-checker.svg?style=flat

[travis-url]: https://travis-ci.org/ivanoff/time-checker
[travis-image]: https://travis-ci.org/ivanoff/time-checker.svg?branch=master

[appveyor-url]: https://ci.appveyor.com/project/ivanoff/time-checker/branch/master
[appveyor-image]: https://ci.appveyor.com/api/projects/status/lp3nhnam1eyyqh33/branch/master?svg=true

[coveralls-url]: https://coveralls.io/github/ivanoff/time-checker?branch=master
[coveralls-image]: https://coveralls.io/repos/github/ivanoff/time-checker/badge.svg?branch=master


CHANGELOG.md
# time-checker changelog

## [0.1.2](https://github.com/ivanoff/time-checker/tree/0.1.2) (2017-01-..)
[Full Changelog](https://github.com/ivanoff/time-checker/compare/0.1.1...0.1.2)

**What Was Done:**

- ...


## [0.1.1](https://github.com/ivanoff/time-checker/tree/0.1.1) (2017-01-12)
[Full Changelog](https://github.com/ivanoff/time-checker/compare/0.0.1...0.1.1)

**What Was Done:**

- add options parameters
- add usage example
- add timezone into options


## [0.0.1](https://github.com/ivanoff/time-checker/tree/0.0.1) (2017-01-11)
[Full Changelog](https://github.com/ivanoff/time-checker/compare/0.0.1...0.0.1)

**What Was Done:**

- made first run
- passed all tests


test/checkTime.js
'use strict';

var CheckTime = require('../index');

var options = {
  daysOff: [ 0, 7 ],
  holidays: [ '1.01', '7.01' ],
  timeZone: 'Etc/GMT'
};

describe('checkTime', function() {

  describe('allways true', function() {
    var checkTime = new CheckTime();
    it('no conditions', function() {
      checkTime.byConditions().should.equal(true);
    });
    it('00:00-23:59', function() {
      checkTime.byConditions('00:00-23:59').should.equal(true);
    });
  });

  describe('allways false', function() {
    var checkTime = new CheckTime();
    it('00:00-00:00', function() {
      checkTime.byConditions('00:00-00:00').should.equal(false);
    });
  });

  describe('1465211606000 => Mon Jun 06 11:13:26 2016', function() {
    var checkTime = new CheckTime(options);
    it('9:00-17:00 string', function() {
      checkTime.byConditions('9:00-17:00', 1465211606000).should.equal(true);
    });
    it('9:00-17:00', function() {
      checkTime.byConditions(['9:00-17:00'], 1465211606000).should.equal(true);
    });
    it('9:00-12:00', function() {
      checkTime.byConditions(['9:00-12:00'], 1465211606000).should.equal(true);
    });
    it('11:00-12:00', function() {
      checkTime.byConditions(['11:00-12:00'], 1465211606000).should.equal(true);
    });
    it('9:00-11:00', function() {
      checkTime.byConditions(['9:00-11:00'], 1465211606000).should.equal(false);
    });
    it('12:00-17:00', function() {
      checkTime.byConditions(['12:00-17:00'], 1465211606000).should.equal(false);
    });
    it('9:00-10:00, 11:00-12:00', function() {
      checkTime.byConditions(['9:00-10:00','11:00-12:00'], 1465211606000).should.equal(true);
    });
    it('9:00-11:00, 12:00-17:00', function() {
      checkTime.byConditions(['9:00-11:00','12:00-17:00'], 1465211606000).should.equal(false);
    });
    it('workdays', function() {
      checkTime.byConditions(['workdays'], 1465211606000).should.equal(true);
    });
    it('11:00-12:00, workdays', function() {
      checkTime.byConditions(['11:00-12:00','workdays'], 1465211606000).should.equal(true);
    });
    it('9:00-11:00, workdays', function() {
      checkTime.byConditions(['9:00-11:00','workdays'], 1465211606000).should.equal(false);
    });
    it('workdays, 11:00-12:00', function() {
      checkTime.byConditions(['workdays','11:00-12:00'], 1465211606000).should.equal(true);
    });
    it('workdays, 9:00-11:00', function() {
      checkTime.byConditions(['workdays','9:00-11:00'], 1465211606000).should.equal(false);
    });
    it('workdays, 11:00-12:00, workdays', function() {
      checkTime.byConditions(['workdays','11:00-12:00','workdays'], 1465211606000).should.equal(true);
    });
    it('workdays, 9:00-11:00, workdays', function() {
      checkTime.byConditions(['workdays','9:00-11:00','workdays'], 1465211606000).should.equal(false);
    });
    it('11:00-12:00, workdays, 11:00-12:00', function() {
      checkTime.byConditions(['11:00-12:00','workdays','11:00-12:00'], 1465211606000).should.equal(true);
    });
    it('9:00-11:00, workdays, 9:00-11:00', function() {
      checkTime.byConditions(['9:00-11:00','workdays','9:00-11:00'], 1465211606000).should.equal(false);
    });
    it('restdays', function() {
      checkTime.byConditions(['restdays'], 1465211606000).should.equal(false);
    });
  });

  describe('1465125226000 => Sun Jun 05 11:13:06 2016', function() {
    var checkTime = new CheckTime(options);
    it('restdays', function() {
      checkTime.byConditions(['restdays'], 1465125226000).should.equal(true);
    });
    it('11:00-12:00, restdays', function() {
      checkTime.byConditions(['11:00-12:00','restdays'], 1465125226000).should.equal(true);
    });
    it('9:00-11:00, restdays', function() {
      checkTime.byConditions(['9:00-11:00','restdays'], 1465125226000).should.equal(false);
    });
    it('restdays, 11:00-12:00', function() {
      checkTime.byConditions(['restdays','11:00-12:00'], 1465125226000).should.equal(true);
    });
    it('restdays, 9:00-11:00', function() {
      checkTime.byConditions(['restdays','9:00-11:00'], 1465125226000).should.equal(false);
    });
    it('restdays, 11:00-12:00, restdays', function() {
      checkTime.byConditions(['restdays','11:00-12:00','restdays'], 1465125226000).should.equal(true);
    });
    it('restdays, 9:00-11:00, restdays', function() {
      checkTime.byConditions(['restdays','9:00-11:00','restdays'], 1465125226000).should.equal(false);
    });
    it('11:00-12:00, restdays, 11:00-12:00', function() {
      checkTime.byConditions(['11:00-12:00','restdays','11:00-12:00'], 1465125226000).should.equal(true);
    });
    it('9:00-11:00, restdays, 9:00-11:00', function() {
      checkTime.byConditions(['9:00-11:00','restdays','9:00-11:00'], 1465125226000).should.equal(false);
    });
  });

  describe('1465125606000 => Sun Jun 05 11:20:06 2016', function() {
    var checkTime = new CheckTime(options);
    it('9:00-17:00', function() {
      checkTime.byConditions(['9:00-17:00'], 1465125606000).should.equal(true);
    });
    it('9:00-12:00', function() {
      checkTime.byConditions(['9:00-12:00'], 1465125606000).should.equal(true);
    });
    it('11:00-12:00', function() {
      checkTime.byConditions(['11:00-12:00'], 1465125606000).should.equal(true);
    });
    it('9:00-11:00', function() {
      checkTime.byConditions(['9:00-11:00'], 1465125606000).should.equal(false);
    });
    it('12:00-17:00', function() {
      checkTime.byConditions(['12:00-17:00'], 1465125606000).should.equal(false);
    });
    it('9:00-10:00, 11:00-12:00', function() {
      checkTime.byConditions(['9:00-10:00','11:00-12:00'], 1465125606000).should.equal(true);
    });
    it('9:00-11:00, 12:00-17:00', function() {
      checkTime.byConditions(['9:00-11:00','12:00-17:00'], 1465125606000).should.equal(false);
    });
    it('workdays', function() {
      checkTime.byConditions(['workdays'], 1465125606000).should.equal(false);
    });
    it('11:00-12:00, workdays', function() {
      checkTime.byConditions(['11:00-12:00','workdays'], 1465125606000).should.equal(false);
    });
    it('9:00-11:00, workdays', function() {
      checkTime.byConditions(['9:00-11:00','workdays'], 1465125606000).should.equal(false);
    });
    it('workdays, 11:00-12:00', function() {
      checkTime.byConditions(['workdays','11:00-12:00'], 1465125606000).should.equal(false);
    });
    it('workdays, 9:00-11:00', function() {
      checkTime.byConditions(['workdays','9:00-11:00'], 1465125606000).should.equal(false);
    });
    it('workdays, 11:00-12:00, workdays', function() {
      checkTime.byConditions(['workdays','11:00-12:00','workdays'], 1465125606000).should.equal(false);
    });
    it('workdays, 9:00-11:00, workdays', function() {
      checkTime.byConditions(['workdays','9:00-11:00','workdays'], 1465125606000).should.equal(false);
    });
    it('11:00-12:00, workdays, 11:00-12:00', function() {
      checkTime.byConditions(['11:00-12:00','workdays','11:00-12:00'], 1465125606000).should.equal(false);
    });
    it('9:00-11:00, workdays, 9:00-11:00', function() {
      checkTime.byConditions(['9:00-11:00','workdays','9:00-11:00'], 1465125606000).should.equal(false);
    });
  });

});

index.js
/***
 * Check current time in a specified period
 * Example of usage:
 *   var CheckTime = require('time-checker');
 *   var checkTime = new CheckTime({ daysOff : [0, 7], holidays : ['1.01', '7.01'] });
 *   // checking 'Mon Jun 06 11:13:26 2016' (1465200806000) on perioud '9:00-17:00'
 *   checkTime.byConditions('9:00-17:00', 1465200806000); // true
 *   // is Sunday is a rest day?
 *   checkTime.byConditions('restdays', 1465200806000); // true
 *   // is 'Mon Jun 06 11:13:26 2016' (1465114426000) is a rest day with perioud '9:00-17:00'?
 *   checkTime.byConditions(['9:00-17:00', 'restdays'], 1465200806000); // false
 *   // if current time is my working days free time?
 *   checkTime.byConditions(['18:00-23:59', '00:00-10:00', 'workdays']); // it depends on current time
 */
'use strict';

var moment = require('moment-timezone');

/**
 * Create a time-checker application
 * @return {Object}
 */
exports = module.exports = function(options) {
  if(!options) options = {};

  if(!options.daysOff) options.daysOff = [];
  if(!options.holidays) options.holidays = [];

  return {
    options: options,
    byConditions : checkTimeByConditions,
  };
}

/**
 * Check whether the current time and date fits under the described conditions.
 * Returns true if the current time and date is within the described terms.
 * Examples:
 *   checkTimeByConditions(['9:00-17:00','workdays'])
 *   checkTimeByConditions(['17:00-23:59','00:00-8:59'])
 */
function checkTimeByConditions(conditions, timestamp) {
  if(!conditions) return true;

  var resultTime = [];
  var resultWeek = true;

  var current = this.options.timeZone? moment(timestamp).tz(this.options.timeZone) : moment(timestamp);

  conditions = Array.isArray(conditions)? conditions : [conditions];
  var conditionsLength = conditions.length;

  for(var i = 0; i < conditionsLength; i++) {
    var e = conditions[i];

    var r = e.match(/^(\d+):(\d+)-(\d+):(\d+)$/);
    if(r && r[4]) {
      var timeFrom = moment(current).hour(r[1]).minute(r[2]);
      var timeTo = moment(current).hour(r[3]).minute(r[4]);
      resultTime.push(current.unix() >= timeFrom.unix() && current.unix() <= timeTo.unix());
    }

    var d = this.options.daysOff;
    var h = this.options.holidays;
    if(e === 'workdays')
      resultWeek = d.indexOf(current.day()) < 0 && h.indexOf(current.format('D.MM')) < 0;
    if(e === 'restdays')
      resultWeek = d.indexOf(current.day()) >= 0 || h.indexOf(current.format('D.MM')) >= 0;

  }

  resultTime = !resultWeek? [] : resultWeek && !resultTime.length? [true] : resultTime;
  return resultTime.indexOf(true) > -1;
}
